0051-options-local-pragma.rst

```diff
@@ -0,0 +1,106 @@
+Options Local Pragma
+====================
+
+.. proposal-number:: 
+.. ticket-url::
+.. implemented::
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/0>`_.
+.. sectnum::
+.. contents::
+
+We propose to add an opportunity to switch GHC-options locally. For instance, this code
+::
+ f = {-# OPTIONS_LOCAL -Wname-shadowing #-}
+   let a = () in
+   let a = 5 in
+     a + a
+get warning ``-Wname-shadowing`` to line
+::
+ let a = 5 in
+
+Motivation
+------------
+
+There are the following reasons to implement such pragma:
+
+1) **Local control of warnings**
+
+To allow control warnings when you confident in your decisions without passing really important warnings.
+
+Warnings are compiler recomendations and often they can be really useful. It allows to avoid stupid mistakes and points to *weak* places in the code. You should follow it. But sometimes you can be confident in your locally code-decision and get some useless warning. It makes tracking warning list more difficult and you can pass really important warning.
+
+ Consider two situations:
+ 
+ a) You want to enable warning locally (and it will be disables in all file).
+ ::
+  f = {-# OPTIONS_LOCAL -Wname-shadowing #-}
+    let a = () in
+    let a = 5 in
+      a + a
+
+ b) You want to disable warning locally (and it will be enabled locally):
+ ::
+  {-# OPTIONS -Wunused-do-bind #-}
+ 
+  main :: IO ()
+  main = {-# OPTIONS_LOCAL -Wno-unused-do-bind #-} do
+    getLine
+    return ()
+
+
+2. **Using language extension locally**
+
+Using language extensions locally to avoid using it in extra places by a misatake.
+
+Language extensions allow to add useful things but "with great power there must also come great responsibility". As a result, sometimes it leads to a lot of mistakes and complicates code.
+
+  You can enable language extension locally (and it will be disabled in all file).
+  ::
+   x' = {-# OPTIONS_LOCAL -XPartialTypeSignatures #-}
+     let a :: _
+         a = ()
+      in ()
+
+
+Proposed Change Specification
+-----------------------------
+
+Work with the GHC options is comfortable throw the special pragma (there is exist pragma ``OPTIONS``). We propose create a similar pragma OPTIONS_LOCAL which will do the same things but locally. You can see how it looks like in the section *Motivation*.
+
+Places for OPTIONS_LOCAL pragma:
+ - expression
+ - bind
+ - declaration
+
+
+Costs and Drawbacks
+-------------------
+
+1) **Estimate on development and maintenance costs**
+
+Every compiler flag can require individual way to collaborate with ``OPTIONS_LOCAL`` pragma.
+
+2) **Influence to learnability of the language**
+
+Similar like work with other pragmas.
+
+3) **Remaining drawbacks**
+
+None.
+
+
+Alternatives
+------------
+
+Pragma `OPTIONS` (but work globally).
```

Comment:
`OPTIONS_GHC` itself it not  an alternative to this proposal. The alternative is “Do nothing”.

But are there really no other alternatives? I vaguely remember that we were discussing something at the very beginning of the project, those options might be worth mentioning in this section.

Reply:
I reread the beginning of the conversation and didn't find any mentions about alternatives.